torch/_refs/nn/functional/__init__.py

```diff
@@ -419,6 +420,122 @@ def hinge_embedding_loss(
     return _apply_loss_reduction(loss, reduction)
 
 
+def _nll_loss_nd(
+    input: TensorLikeType,
+    target: TensorLikeType,
+    weight: Optional[TensorLikeType],
+    reduction: str,
+    ignore_index: int,
+) -> TensorLikeType:
+    if input.ndim == 3 or input.ndim > 4:
+        msg = "Expected input dimension to be either [1, 2, 4] but recieved {}."
+        raise ValueError(msg.format(input.ndim))
+
+    if input.ndim != 1 and input.shape[0] != target.shape[0]:
+        msg = "Expected input batch size ({}) to match target batch size ({})."
+        raise ValueError(msg.format(input.shape[0], target.shape[0]))
+
+    _check_reduction_value(reduction)
+
+    flat_target = torch.reshape(target, [-1])
+    ignore_classes_mask = torch.eq(flat_target, ignore_index)
+    ignore_class_weight = torch.scalar_tensor(0, dtype=input.dtype, device=input.device)
+    default_class_weight = torch.scalar_tensor(
+        1, dtype=input.dtype, device=input.device
+    )
+
+    # TODO: This check does not work with FakeTensor inputs
+    """
+    num_classes = input.shape[1] if input.ndim > 1 else input.shape[0]
+    valid_classes_mask = torch.logical_and((flat_target >= 0), (flat_target < num_classes))
+    if not torch.all(torch.logical_or(ignore_classes_mask, valid_classes_mask)):
+        print(target, num_classes, ignore_index)
+        raise ValueError("Target class is out-of-bounds and not ignore index")
+    """
+
+    if weight is None:
```

Comment:
This will need a comment, particularly to explain the expansion.

Instead of branching here it might be better to prepare the arguments before calling `where`